keps/sig-architecture/4402-go-workspaces/README.md

```diff
@@ -0,0 +1,414 @@
+understood GOPATH, and it leaked into many parts of our build, test, and
+release systems.  Then Go added modules, in part because GOPATH was unpleasant
+and had a tendency to leak into tools.  We adopted modules, but we also added
+the idea of "staging" repositories - a way to eat our cake and have, too.  We
+get the benefits of a monorepo (atomic commits, fast iteration across repos)
+and then publish those to standalone repos for downstream consumption.  To make
+this work with modules, we abused GOPATH even harder and wrote even more tools.
+
+The Go project saw what we (and others) were doing and did not like it.  So
+they created [workspaces](https://go.dev/doc/tutorial/workspaces).  They
+basically created a solution that is purpose-built for us.
+
+This KEP proposes that we adopt Go workspaces and bring our tooling up to
+modern standards.  In fact, the author started this work in 2021 or 2022,
+discovering issues along the way that the Go team has worked to resolve.
+
+This is not a user-facing change.  No k8s cluster-user or cluster-admin should
+know or care that this happened.  On the surface, it seems like something that
+should not warrant a KEP, but:
+  a) KEPs are how we communicate big changes;
+  b) This is a big change;
+  c) There's some ecosystem impact to developers
+
+## Motivation
+
+Anyone who has had to maintain the multitude of `verify-*` and `update-*`
+scripts has felt this pain.  Anyone who has had to deal with how we build and
+test Kubernetes has run into the mess.  It's a constant source of complexity
+and friction for maintainers, and it leaks out into our developer ecosystem.
+
+This is the sort of mess that wise people back away from.  The author is not so
+wise.
+
+### Goals
+
+1) To get rid of our reliance on GOPATH hackery to build Kubernetes and to run
+   verifier tools and code-generators.
+2) To simplify and clarify our tooling with regards to multi-module operation.
+3) To be able to use upstream tooling, such as `gopls` across multiple modules.
+
+### Non-Goals
+
+1) To rewrite our code-generators (though some could use it).
+2) To make the build or tools faster.
+
+## Proposal
+
+This KEP proposes to add a `go.work` file (checked in) and fix all the tools
+that break because of that.  At the end, there will be no reliance on GOPATH.
+verify and update scripts can use normal Go conventions for building and
+running code.  Abominations like `run-in-gopath.sh` will be deleted.
+
+The implementation will come in several waves, which will have to be sequenced
+across k8s.io/kubernetes, k8s.io/gengo, and k8s.io/kube-openapi repos.  In
+order to not leave repos in a broken state, some PRs will need to be larger
+than we might otherwise like, but commits can tell the story.
+
+This is a proposed merge sequence (though the development PR is in a different
+order).
+
+1. Retool the core of gengo to handle workspaces.  This will be gengo/v2.  The
```

Comment:
The decision process that arrived at `gengo/v2` was interesting to me (as was underlying topic of how careful we should be about publishing libraries for general purpose tasks like codegen).  Anything about that worth capturing in the KEP?

Reply:
Sure, I can do that